stan/math/rev/functor/adj_arg.hpp

```diff
@@ -11,6 +11,57 @@
 namespace stan {
 namespace math {
 
+template <typename T, typename = void>
+struct adj_argr;
+
+template <typename T>
+struct adj_argr<T, require_container_t<T>> {
+  using FunctorReturnType = plain_type_t<decltype(value_of(std::declval<T>()))>;
+  // If the type is an `std::vector` this stores an `Eigen::VectorXd`
+  using RetType
+      = std::conditional_t<is_std_vector<FunctorReturnType>::value,
+                           Eigen::Matrix<double, -1, 1>, FunctorReturnType>;
+  // The Eigen map type used for storage
+  using type = Eigen::Map<RetType>;
+};
+
+template <typename T>
+struct adj_argr<T, require_stan_scalar_t<T>> {
+  using type = double;
+};
+
+template <typename T>
+using adj_arg_t = typename adj_argr<std::decay_t<T>>::type;
+
+template <typename T, bool SaveValue = true>
+auto setup_adj_arg(size_t n) {
+  double* mem = ChainableStack::instance_->memalloc_.alloc_array<double>(n);
+  return adj_arg_t<T>(mem, n);
+}
+
+template <typename T, bool SaveValue,
+          require_t<bool_constant<!SaveValue>>* = nullptr>
```

Comment:
```suggestion
          std::enable_if_t<!SaveValue>* = nullptr>
```

Reply:
That SaveValue is confusing me a little. For some arguments there exists only an overload that accepts SaveValue=flase and none for SaveValue=true, and for others SaveValue is ignored.